Improve routes management

jmaupetit opened the pull request:
## Purpose

Listing routes from the `apps` global variable definition was not that explicit nor easy to maintain. Hence, following our global "app-driven" logic, we decided to manage routes from their definition stated in an application service template. 

## Proposal

- [x] upgrade routes discovery and creation
- [x] remove hosts/ports definitions in the `apps` global var
- [x] add routes for every application service that requires one or more

As a consequence, this PR will fix #90.

jmaupetit committed:
🐛(routes) fix redirect route template path

Since the redirect application tree refactoring, it seems that the path
to the route generic template did not follow. This is now fixed.

jmaupetit committed:
⚡️(routes) discover routes from app services

Listing routes from the apps global variable definition was not that
explicit nor easy to maintain. Hence, following our global app-driven
logic, we decided to manage routes from their definition stated in an
application service template. A route template file should be located in
the ad hoc service and its filename should match the following pattern:
`route.*\.yml\.j2`.

jmaupetit committed:
✨(routes) add application services route templates

Now that routes should be explicitly defined as a template for each
service they refer to, we've added them for richie and edxapp (cms+lms).

sampaccoud commented on apps/richie/templates/nginx/route.yml.j2:
```diff
@@ -0,0 +1,43 @@
+    Host prefix can be: previous, next or current. You can use this template with
+    either a prefix set using the following variables (in precedence order):
+
+        - item.1 (in a nested loop)
```
I don't understand "in a nested loop"...

sampaccoud replied:
can item be a namedtuple of a dictionary so we write something else than "item.1" !? Can item have a better name also ? it's an item of what?

jmaupetit replied:
I'm referring to Ansible nested loops: https://docs.ansible.com/ansible/2.5/user_guide/playbooks_loops.html#complex-loops

But I got your point and I'll come with a better idea / implementation.

jmaupetit committed:
✨(plugins) add blue_green_host filter

Distributed templates should be the simpler they can be. It means that
we need to move most the underlying logic out of them. With this
philosophy in mind, we've created a small filter that adds the
blue-green prefix to a base host.

jmaupetit committed:
⚡️(apps) make the hello app suite new app tree

The hello application was not respecting new standars and expected
files. This is now fixed.

jmaupetit committed:
🔥(vars) clean up legacy apps metadata & unused vars

We removed no longer required applications metadata used for routes and
cleaned up useless overriden vars.

cchaudier commented on group_vars/customer/hello/main.yml:
```diff
@@ -6,14 +6,6 @@ project_description: "A minimal project with a hello-world app ({{ env_type }})"
-hello_port: "8080"
-hello_msg:  "Hello OpenShift! by Arnold"
 
 nginx_ports:
```
This also  soon deprecated

cchaudier commented on group_vars/all/main.yml:
```diff
@@ -34,6 +34,12 @@ endpoint_mongodb_ip:     192.168.10.11
+blue_green_route_prefixes:
+  - "previous"
+  - "current"
+  - "next"
```
I'v do this before but 
@sampaccoud on 12 Jun  ask this :
I'm skeptical that this should be configurable. Keep it simple and impose it?

See : https://github.com/openfun/arnold/pull/60

jmaupetit replied:
:thinking: it might evolve...

cchaudier replied:
Probably for me it's great like this ^^

cchaudier commented on plugins/filter/deploy.py:
```diff
@@ -0,0 +1,46 @@
+    def filters(self):
+        """List plugin filters"""
+
+        return {"blue_green_host": blue_green_host}
```
This  filter is just for not using the nested loop ?

jmaupetit replied:
No this is to avoid adding logic to the template.

cchaudier replied:
Ok in this case

cchaudier commented on apps/edxapp/templates/nginx/route_cms.yml.j2:
```diff
@@ -0,0 +1,25 @@
+  annotations:
+    kubernetes.io/tls-acme: "true"
+spec:
+  host: "{{ edxapp_cms_host | blue_green_host(prefix) }}"
```
This this the host can't be overrided in the apps dict.
If whe want otherride this host we must declare `edxapp_cms_host` in the group_vars.

jmaupetit replied:
>  If whe want otherride this host we must declare edxapp_cms_host in the group_vars.

Yes.

cchaudier replied:
Ok but I think the apps dict lose is first interest after your PR.

jmaupetit replied:
I disagree and you will see why in the next PR.

cchaudier reviewed (commented):
Huge job ! :+1: 
I've somme comments
